File: src/collections_and_containers/cpp/avl_tree/src/avl_tree.cpp
#include "avl_tree.h"
#include <algorithm>
#include <stdexcept>

template <class T> AvlTree<T>::AvlTree() : root(nullptr), n(0) {}

template <class T>
AvlTree<T>::AvlTree(const AvlTree<T> &t) : root(nullptr), n(0) {
  for (const auto &v : t.in_order_traversal()) {
    insert(v);
  }
}

template <class T> AvlTree<T> &AvlTree<T>::operator=(const AvlTree<T> &t) {
  if (this != &t) {
    clear();
    for (const auto &v : t.in_order_traversal()) {
      insert(v);
    }
  }
  return *this;
}

template <class T> AvlTree<T>::~AvlTree() { clear(); }

template <class T> void AvlTree<T>::insert(const T &v) { insert(v, root); }

template <class T> void AvlTree<T>::remove(const T &v) { remove(v, root); }

template <class T> bool AvlTree<T>::contains(const T &v) const {
  return contains(v, root);
}

template <class T> T AvlTree<T>::find_min() const {
  if (root == nullptr) {
    throw std::out_of_range("Tree is empty");
  }
  const Node *p = root.get();
  while (p->left != nullptr) {
    p = p->left.get();
  }
  return p->data;
}

template <class T> T AvlTree<T>::find_max() const {
  if (root == nullptr) {
    throw std::out_of_range("Tree is empty");
  }
  const Node *p = root.get();
  while (p->right != nullptr) {
    p = p->right.get();
  }
  return p->data;
}

template <class T> void AvlTree<T>::clear() { clear(root); }

template <class T> int AvlTree<T>::height() const { return height(root); }

template <class T> int AvlTree<T>::size() const { return n; }

template <class T> bool AvlTree<T>::empty() const { return n == 0; }

template <class T> std::vector<T> AvlTree<T>::in_order_traversal() const {
  std::vector<T> v;
  in_order_traversal(root, v);
  return v;
}

template <class T> std::vector<T> AvlTree<T>::pre_order_traversal() const {
  std::vector<T> v;
  pre_order_traversal(root, v);
  return v;
}

template <class T> std::vector<T> AvlTree<T>::post_order_traversal() const {
  std::vector<T> v;
  post_order_traversal(root, v);
  return v;
}

template <class T>
void AvlTree<T>::insert(const T &v, std::unique_ptr<Node> &p) {
  if (p == nullptr) {
    p = std::make_unique<Node>();
    p->data = v;
    n++;
  } else if (v < p->data) {
    insert(v, p->left);
  } else if (v > p->data) {
    insert(v, p->right);
  } else {
    throw std::invalid_argument("duplicate value");
  }
  balance(p);
}

template <class T>
void AvlTree<T>::remove(const T &v, std::unique_ptr<Node> &p) {
  if (p == nullptr) {
    throw std::invalid_argument("value not found");
  } else if (v < p->data) {
    remove(v, p->left);
  } else if (v > p->data) {
    remove(v, p->right);
  } else {
    if (p->left == nullptr && p->right == nullptr) {
      p = nullptr;
      n--;
    } else if (p->left == nullptr) {
      p = std::move(p->right);
      n--;
    } else if (p->right == nullptr) {
      p = std::move(p->left);
      n--;
    } else {
      // Find in-order predecessor
      Node *pred = p->left.get();
      while (pred->right != nullptr) {
        pred = pred->right.get();
      }
      p->data = pred->data;
      remove(pred->data, p->left);
    }
  }
  if (p != nullptr) {
    balance(p);
  }
}

template <class T>
bool AvlTree<T>::contains(const T &v, const std::unique_ptr<Node> &p) const {
  if (p == nullptr) {
    return false;
  } else if (v < p->data) {
    return contains(v, p->left);
  } else if (v > p->data) {
    return contains(v, p->right);
  } else {
    return true;
  }
}

template <class T> void AvlTree<T>::clear(std::unique_ptr<Node> &p) {
  if (p != nullptr) {
    clear(p->left);
    clear(p->right);
    p = nullptr;
    n--;
  }
}

template <class T> int AvlTree<T>::height(const std::unique_ptr<Node> &p) const {
  if (p == nullptr) {
    return 0;
  } else {
    return 1 + std::max(height(p->left), height(p->right));
  }
}

template <class T>
void AvlTree<T>::in_order_traversal(const std::unique_ptr<Node> &p,
                                    std::vector<T> &v) const {
  if (p != nullptr) {
    in_order_traversal(p->left, v);
    v.push_back(p->data);
    in_order_traversal(p->right, v);
  }
}

template <class T>
void AvlTree<T>::pre_order_traversal(const std::unique_ptr<Node> &p,
                                     std::vector<T> &v) const {
  if (p != nullptr) {
    v.push_back(p->data);
    pre_order_traversal(p->left, v);
    pre_order_traversal(p->right, v);
  }
}

template <class T>
void AvlTree<T>::post_order_traversal(const std::unique_ptr<Node> &p,
                                      std::vector<T> &v) const {
  if (p != nullptr) {
    post_order_traversal(p->left, v);
    post_order_traversal(p->right, v);
    v.push_back(p->data);
  }
}

template <class T> void AvlTree<T>::balance(std::unique_ptr<Node> &p) {
  if (p == nullptr) {
    return;
  }
  
  int leftHeight = height(p->left);
  int rightHeight = height(p->right);
  
  if (leftHeight - rightHeight > 1) {
    // Left-heavy
    if (height(p->left->left) >= height(p->left->right)) {
      // Left-Left case
      rotate_right(p);
    } else {
      // Left-Right case
      rotate_left(p->left);
      rotate_right(p);
    }
  } else if (rightHeight - leftHeight > 1) {
    // Right-heavy
    if (height(p->right->right) >= height(p->right->left)) {
      // Right-Right case
      rotate_left(p);
    } else {
      // Right-Left case
      rotate_right(p->right);
      rotate_left(p);
    }
  }
}

template <class T> void AvlTree<T>::rotate_left(std::unique_ptr<Node> &p) {
  std::unique_ptr<Node> r = std::move(p->right);
  p->right = std::move(r->left);
  r->left = std::move(p);
  p = std::move(r);
}

template <class T> void AvlTree<T>::rotate_right(std::unique_ptr<Node> &p) {
  std::unique_ptr<Node> l = std::move(p->left);
  p->left = std::move(l->right);
  l->right = std::move(p);
  p = std::move(l);
}

template class AvlTree<int>;
template class AvlTree<std::string>;
template class AvlTree<double>;
template class AvlTree<float>;

File: src/collections_and_containers/python/avl_tree/src/avl_tree.py
"""AVL Tree (self-balancing binary search tree) implementation."""

from __future__ import annotations

from typing import Any, Optional


class Node:
    """A node in an AVL tree."""

    def __init__(self, value: Any) -> None:
        self.value = value
        self.left: Optional[Node] = None
        self.right: Optional[Node] = None
        self.balance_factor = 0


class AvlTree:
    """A self-balancing AVL tree data structure."""

    def __init__(self) -> None:
        self.root: Optional[Node] = None
        self.n = 0

    def insert(self, value: Any) -> None:
        """Insert a value into the tree."""
        def insert_helper(p: Optional[Node]) -> Node:
            if not p:
                return Node(value)

            if value < p.value:
                p.left = insert_helper(p.left)
            else:
                p.right = insert_helper(p.right)

            right_height = self._height(p.right) if p.right else 0
            left_height = self._height(p.left) if p.left else 0
            p.balance_factor = right_height - left_height
            return self.balance(p)

        self.root = insert_helper(self.root)
        self.n += 1

    def remove(self, value: Any) -> None:
        """Remove a value from the tree."""
        def remove_helper(p: Optional[Node]) -> Optional[Node]:
            if not p:
                return None

            if value == p.value:
                if not p.left:
                    return p.right
                if not p.right:
                    return p.left

                min_value = self.find_min(p.right)
                p.value = min_value
                p.right = remove_helper(p.right)
            elif value < p.value:
                p.left = remove_helper(p.left)
            else:
                p.right = remove_helper(p.right)

            right_height = self._height(p.right) if p.right else 0
            left_height = self._height(p.left) if p.left else 0
            p.balance_factor = right_height - left_height
            return self.balance(p)

        self.root = remove_helper(self.root)
        self.n -= 1

    def contains(self, value: Any) -> bool:
        """Check if the tree contains a value."""
        def contains_helper(p: Optional[Node]) -> bool:
            if not p:
                return False
            if value == p.value:
                return True
            elif value < p.value:
                return contains_helper(p.left)
            else:
                return contains_helper(p.right)

        return contains_helper(self.root)

    def clear(self) -> None:
        """Remove all elements from the tree."""
        self.root = None
        self.n = 0

    def height(self) -> int:
        """Return the height of the tree."""
        return self._height(self.root)

    def _height(self, p: Optional[Node]) -> int:
        if not p:
            return 0
        return 1 + max(self._height(p.left), self._height(p.right))

    def size(self) -> int:
        """Return the number of elements in the tree."""
        return self.n

    def empty(self) -> bool:
        """Check if the tree is empty."""
        return self.n == 0

    def in_order_traversal(self) -> list[Any]:
        """Return elements in in-order traversal (left, root, right)."""
        def in_order_helper(p: Optional[Node], v: list[Any]) -> None:
            if p:
                in_order_helper(p.left, v)
                v.append(p.value)
                in_order_helper(p.right, v)

        v: list[Any] = []
        in_order_helper(self.root, v)
        return v

    def pre_order_traversal(self) -> list[Any]:
        """Return elements in pre-order traversal (root, left, right)."""
        def pre_order_helper(p: Optional[Node], v: list[Any]) -> None:
            if p:
                v.append(p.value)
                pre_order_helper(p.left, v)
                pre_order_helper(p.right, v)

        v: list[Any] = []
        pre_order_helper(self.root, v)
        return v

    def post_order_traversal(self) -> list[Any]:
        """Return elements in post-order traversal (left, right, root)."""
        def post_order_helper(p: Optional[Node], v: list[Any]) -> None:
            if p:
                post_order_helper(p.left, v)
                post_order_helper(p.right, v)
                v.append(p.value)

        v: list[Any] = []
        post_order_helper(self.root, v)
        return v

    def balance(self, p: Optional[Node]) -> Optional[Node]:
        """Balance the tree at node p."""
        if not p:
            return None

        if p.balance_factor == 2:
            if not p.right:
                return p
            if p.right.balance_factor < 0:
                p.right = self.rotate_right(p.right)
            return self.rotate_left(p)
        elif p.balance_factor == -2:
            if not p.left:
                return p
            if p.left.balance_factor > 0:
                p.left = self.rotate_left(p.left)
            return self.rotate_right(p)
        return p

    def rotate_left(self, p: Node) -> Node:
        """Perform a left rotation."""
        q = p.right
        p.right = q.left
        q.left = p
        return q

    def rotate_right(self, p: Node) -> Node:
        """Perform a right rotation."""
        q = p.left
        p.left = q.right
        q.right = p
        return q

    def find_min(self, p: Node) -> Any:
        """Find the minimum value in the subtree rooted at p."""
        while p.left:
            p = p.left
        return p.value
